### PR TITLE
[Core] Optimize the memory footprint of `owner_address` field with boost::flyweight

### PR DESCRIPTION
**Pull request by NKcqx**
## Why are these changes needed?
This PR aims to reduce the metadata overhead that has been described in #28765 .
One of the overhead comes from the `owner_address` field of object's reference. The `owner_address` has large redundancy because there are only few owners in cluster, but their `rpc::Address` are saved per object. Therefore we can use an intern table to maintain the `rpc::Address`.

> boost::flyweight provide an implementation of the intern table that gives access to shared common data. [docs](https://www.boost.org/doc/libs/1_77_0/libs/flyweight/doc/index.html).

### Experiment
To reduce the memory overhead, we need to first profile and analyze the runtime memory footprint to locate the problems.
To profile and analyze:  _jemalloc_ to profile the memory usage and  _jeprof_ to visualize;
To get end2end memory footprint: see test code in #28765 .

#### Setup
jemalloc: We custom build the jemalloc to [enable memory profiling ability](https://docs.percona.com/percona-server/8.0/performance/jemalloc-profiling.html#:~:text=To%20enable%20jemalloc%20profiling%20in,the%20results%20as%20a%20table).

jeprof: `yum install ghostscript graphviz jemalloc-devel -y`

> To get a more accurate profiling result, we set a smaller profiling interval, i.e. `lg_prof_interval`, to about 25~28, and highest sample frequency, i.e. `lg_prof_sample:0` . 
Note that, this will significantly lower the execution speed and create a large number of heap file fragments, be care for the disk usage.

#### scripts:
```bash
#!/usr/bin/env bash

set -e

yum install autoconf bzip2 -y

pushd /home/admin/tools
wget https://github.com/jemalloc/jemalloc/releases/download/5.3.0/jemalloc-5.3.0.tar.bz2
tar -xf jemalloc-5.3.0.tar.bz2
pushd jemalloc-5.3.0
./configure --enable-prof
make
make install
popd
rm -rf jemalloc-5.3.0*
popd

JEMALLOC_FILE=/usr/local/lib/libjemalloc.so

yum install ghostscript graphviz jemalloc-devel -y

log_path="/tmp/jemalloc/dumps"

mkdir -p $log_path
# jemalloc config
export MALLOC_CONF="prof:true,prof_accum:true,lg_prof_interval:28,prof_prefix:$log_path/jeprof.out"
export MALLOC_CONF="$MALLOC_CONF,lg_prof_sample:0"
# start exp
OBJ_CNT=10000
LD_PRELOAD="$JEMALLOC_FILE" \
python -c  '
import ray
import time
import sys
import os, psutil
import matplotlib.pyplot as plt

from tqdm import tqdm

process = psutil.Process(os.getpid())

ray.init()

@ray.remote
class Owner:
    def foo(self):
        return 1
o = Owner.remote()
assert ray.get(o.foo.remote())

ref = ray.put("data")

@ray.remote
class A:
    def warmup(self, owner, batch_size=1):
        return [ray.put("data", _owner=owner) for _ in range(batch_size)]
a = A.remote()
assert ray.get(a.warmup.remote(o))

size = 10000
batch_size = 100
vec = [None] * size
interval = int(size * 0.02)
rss_list = []
x_axis = [ i * interval for i in range(int(size / interval))]


print("==== start ====")
for i in tqdm(range(size)):
    if  i % interval == 0:
        mem = process.memory_info().rss / 1024 / 1024
        rss_list.append(mem)
        print(f"==== iter {i}: memory usage {mem} MB ====")
    vec[i] = ray.get(a.warmup.remote(o))
' &
# get driver pid
pid=$!
wait $pid;
file_prefix="$log_path"/jeprof.out."$pid"
base_snapshot=$(ls -d "$file_prefix"* | head -n 1)
final_snapshot=$(ls -d "$file_prefix"* | tail -n 1)
# draw
jeprof --svg --inuse_objects --lines --nodecount=10000 --base="$base_snapshot" `which python`  "$final_snapshot"  > "$log_path"/"$pid"-put"$OBJ_CNT".count.svg
jeprof --svg --inuse_space --show_bytes --lines --nodecount=10000 --base="$base_snapshot" `which python`  "$final_snapshot"  > "$log_path"/"$pid"-put"$OBJ_CNT".memory.svg
rm -rf /tmp/jemalloc/dumps/*.heap
```

#### Result

The e2e result, i.e. the rss reported by `psutil.Process.memory_info()` , is influenced by many runtime factors, such as the number of objects will influence the load factor of used hashset & hashmap; A low load factor will then result in higher memory footprint per object since there are more waste of space.
Therefore only a fluctuate range can be given.

Before optimize, the average rss is about 2240 bytes;
After optimize, the average rss is about 1933 bytes.
See the  [two example running case](https://gist.github.com/NKcqx/68b0424245453d35d92da52bd38ad786).
 
**Therefore, this optimization can save about 310 bytes.**

## Related issue number
One optimization for  #28765 .

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/reference_count.h
Comment (kfstorm) [inline src/ray/core_worker/reference_count.h:39]:
```suggestion
struct AddressWorkerIdExtractor {
```

**Review by kfstorm — APPROVED** [review]
LGTM.

**Comment by kfstorm** [comment]
@stephanie-wang Could you take some time to review this PR?

**Comment by wumuzi520** [comment]
> Thanks for the optimization. Seems to me boost/flyweight is not widely used when I survey the github, and it's functionality could be easily vendored by ourself; have you considered implemented it without depending on flyweight? also cc @iycheng for thought.

Yes, boost::flyweight is rarely used in other project, but it is do quite suitable to reduce the memory overhead caused by those repeated bytes, and I think it will be complex to rewrite a new class to replace it.

File: thirdparty/patches/rules_boost-flyweight.patch
Comment (jjyao) [inline thirdparty/patches/rules_boost-flyweight.patch:1]: Why do we need this patch?

Comment (NKcqx) [reply thirdparty/patches/rules_boost-flyweight.patch:1]: bcz the boost bazel version we used, i.e. "https://github.com/nelhage/rules_boost/archive/652b21e35e4eeed5579e696da0facbe8dba52b1f.tar.gz", doesn't include the flyweight module.

**Comment by NKcqx** [comment]
> > Thanks for the optimization. Seems to me boost/flyweight is not widely used when I survey the github, and it's functionality could be easily vendored by ourself; have you considered implemented it without depending on flyweight? also cc @iycheng for thought.
> 
> Yes, boost::flyweight is rarely used in other project, but it is do quite suitable to reduce the memory overhead caused by those repeated bytes, and I think it will be complex to rewrite a new class to replace it.

I Agree. Implementing a thread-safe intern table like the flyweight from scratch not only has to write a lot more codes compared to this optimization, but also may incur more unstable factors than a boost module.

**Comment by ericl** [comment]
The flyweight pattern seems like something we can adopt throughout Ray to save memory. It seems easy enough to replace in the future too, if we really want to customize our own intern table.

**Comment by NKcqx** [comment]
> The flyweight pattern seems like something we can adopt throughout Ray to save memory. It seems easy enough to replace in the future too, if we really want to customize our own intern table.

Means that we may customized an intern table later when the motivations are strong ?  I'm willing to carry on if needed.

What do we do on this optimization ?

**Review by wumuzi520 — APPROVED** [review]
LGTM!

**Comment by NKcqx** [comment]
@ericl @jjyao Any other suggestions ?

**Comment by NKcqx** [comment]
The CI failures are unrelated.

File: thirdparty/patches/rules_boost-flyweight.patch
Comment (NKcqx) [reply thirdparty/patches/rules_boost-flyweight.patch:1]: oh, by saying boost version I meant the `rules_boost`. I had tried update the `rule_boost` to the version that includes the `flyweight`, but that version has conflicts with our `windows-linkopts` patch. I was worried that a newer `rule_boost` may further bring more latent problems.

Comment (jjyao) [reply thirdparty/patches/rules_boost-flyweight.patch:1]: Do you mind spending some time to fix the `windows-linkopts` patch? I think eventually we need to be able to update the `rule_boost` version.

We can make sure the upgrade doesn't break things by looking at CI results.

**Comment by NKcqx** [comment]
@jjyao After working on upgrading `rules_boost` for a while, I found that there are a lot follow up problems to be resolved.  Before the `rules_boost`'s  flyweight [commit](https://github.com/nelhage/rules_boost/commit/03876503436f087427fa206a8909b6109a5291e0), a [commit](https://github.com/nelhage/rules_boost/commit/5ce04900fa8f146ddde81cf59cf13fd1b7aca867) has upgrade the boost version to 1.79.0, so we had to upgrading our boost version from 1.74.0 to 1.79.0 and here comes the problems.

[Windows10 starts to support AF_UNIX](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/), i.e. domain socket, and therefore boost enables this feature in 1.79.0. By comparing[ the new version](https://github.com/boostorg/asio/blame/32f8f72fcc77fb88a583875fcefe4ed11528b826/include/boost/asio/detail/config.hpp#L1736) to the [old version](https://github.com/boostorg/asio/blob/boost-1.74.0/include/boost/asio/detail/config.hpp), a key difference is that in `config.hpp` boost starts to set the `BOOST_ASIO_HAS_LOCAL_SOCKETS` to `1` in Win10 environment.

**Comment by jovany-wang** [comment]
@jjyao I think we can merge this PR first. For upgrading boost, we don't have the windows environment. If you guys have the resources, I believe it's better that you guys take the upgrading.

File: src/ray/util/util.cc
Comment (jovany-wang) [inline src/ray/util/util.cc:41]: Remove these debug messages.

File: src/ray/util/util.cc
Comment (jovany-wang) [inline src/ray/util/util.cc:123]: This line should be changed to:
```c++
#if BOOST_ASIO_HAS_LOCAL_SOCKETS
```
instead of `#ifdef`.
So you can remove `!defined(_WIN32)` and `defined(xxx)`.

I believe  it's a bug indeed before this PR, but the lower version boost couldn't trigger it.
Image that if `BOOST_ASIO_HAS_LOCAL_SOCKETS` was set to 0, the if block still be executed.

File: src/ray/core_worker/reference_count.h
Comment (jovany-wang) [inline src/ray/core_worker/reference_count.h:37]:
```suggestion
/// Extract a key to identify an Address so that the flyweight can look
```

File: src/ray/core_worker/reference_count.h
Comment (jovany-wang) [inline src/ray/core_worker/reference_count.h:48]: `using` is a modern keyword than typedef?

**Comment by jovany-wang** [comment]
> @NKcqx, I saw you latest commit upgrade boost to 1.79, does that mean the issue is solved now.

It's a Ray bug to make we couldn't use a high version. The bug trigger some code path which is not allowed on windows.

**Comment by jjyao** [comment]
So the latest commit bypass the issue by adding the additional macro check right?

**Comment by jovany-wang** [comment]
> So the latest commit bypass the issue by adding the additional macro check right?

Not exactly, I believe it's a fix instead of bypass. But it should be changed to the following line as a better fix:
```diff
- #if !defined(_WIN32) && defined(BOOST_ASIO_HAS_LOCAL_SOCKETS)
+ if BOOST_ASIO_HAS_LOCAL_SOCKETS
```

**Comment by jovany-wang** [comment]
I'll take a look again before merging this PR.

**Review by wumuzi520 — APPROVED** [review]

**Comment by jovany-wang** [comment]
@jjyao @simon-mo Do you know what's the version of our windows CI ?

File: src/ray/util/util.cc
Comment (jovany-wang) [inline src/ray/util/util.cc:86]: I guessed a wrong value of this flag. It's always set to 1 on windows CI.

Due to the CI error message shows `sockaddr_un` is not defined, I believe we should add some including lines for it:
image

I'm testing it, I believe it should be pass.

**Review by jovany-wang — CHANGES_REQUESTED** [review]
Let's hold until the CI gets green.

**Comment by NKcqx** [comment]
> Apply this patch to fix the building issue.
> 
> ```diff
> commit 5888a31130f027df9b46e496fafd114c3725c5d7 (HEAD -> flyweight_address)
> Author: Qing Wang <kingchin1218@gmail.com>
> Date:   Sat Oct 29 00:42:27 2022 +0800
> 
>     Fix using boost afunix on windows
> 
>     Signed-off-by: Qing Wang <kingchin1218@gmail.com>
> 
> diff --git src/ray/util/util.cc src/ray/util/util.cc
> index 8f7d41341..bd5e2e2b8 100644
> --- src/ray/util/util.cc
> +++ src/ray/util/util.cc
> @@ -16,7 +16,13 @@
> 
>  #include <stdio.h>
>  #include <stdlib.h>
> -#ifndef _WIN32
> +
> +#ifdef _WIN32
> +#include <windows.h>
> +#include <winsock2.h>
> +#include <ws2tcpip.h>
> +#include <afunix.h>
> +#else
>  #include <sys/un.h>
>  #endif
> 
> @@ -25,9 +31,7 @@
>  #include <sstream>
>  #include <string>
>  #include <vector>
> -#ifndef _WIN32
>  #include <boost/asio/local/stream_protocol.hpp>
> -#endif
>  #include <boost/asio/ip/tcp.hpp>
> 
>  #include "ray/util/filesystem.h"
> ```

It works ~ Really thx for that !

**Comment by jjyao** [comment]
Seems some tests are failing.

**Comment by NKcqx** [comment]
> Seems some tests are failing.

Yeah, win CI start to fail again after moving header positions :(
Working on it.

**Comment by jjyao** [comment]
Thanks a lot for pushing on this. Really glad that we are able to upgrade boost!!

File: thirdparty/patches/boost-exception-no_warn_typeid_evaluated.patch
Comment (jovany-wang) [inline thirdparty/patches/boost-exception-no_warn_typeid_evaluated.patch:1]: Why we remove this file? Is it out of date in the new version?

**Review by jovany-wang — APPROVED** [review]
Code change looks good to me now, but the CI seems to be related.

File: thirdparty/patches/boost-exception-no_warn_typeid_evaluated.patch
Comment (NKcqx) [reply thirdparty/patches/boost-exception-no_warn_typeid_evaluated.patch:1]: Yes, there's no such assertion anymore in [boost 1.79.0](https://github.com/boostorg/exception/blob/boost-1.79.0/include/boost/exception/info.hpp),  and I've search the usage of `BOOST_EXCEPTION_DYNAMIC_TYPEID` to ensure that it's not just been moved to other places.

**Comment by NKcqx** [comment]
Found the preliminary conclusions of the CI failures, which is also commented in codes:
Currently, somewhere exists a reference leak problem, i.e. some processes still hold the reference table instance after Ray shutdown, which prevents the references being freed from _flyweight_. In the some UTs, since the driver process won't exit after each test case, but restart a new Ray cluster, the UT may fail because the flyweight still holds the information of the old Ray cluster and pass it to the new cluster.

The current solution is temporally clear the reference table when core_worker exit. In the future, we may need to find out where the leakage comes from.
@jovany-wang  @jjyao , cc @Catch-Bull

**Comment by iycheng** [comment]
Flyweight is not a very complicated libs. I feel probably it's ok to use it here. We can see whether it brings any problems or not and decide whether to apply this to other components or not.

**Review by jjyao — APPROVED** [review]
Could you file a github issue about the reference table leak? In the issue, could you provide a minimal reproduction script?

**Comment by NKcqx** [comment]
> Could you file a github issue about the reference table leak? In the issue, could you provide a minimal reproduction script?

Sure. Posted in [here](https://github.com/ray-project/ray/issues/30089).

File: src/ray/core_worker/core_worker.cc
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:572]: Actually is there a way to clear the flyweight pool?

Comment (NKcqx) [reply src/ray/core_worker/core_worker.cc:572]: I'm afraid not.

> The central repository of shared value objects is known as the flyweight factory. This component is able to locate and return a reference to an object with a given value, or insert the value if no copy was previously stored. Boost.Flyweight controls the interaction of flyweights with their factory transparently to the programmer, so that a casual user of the library need not even be concerned about the presence of such factory.   [Doc](https://www.boost.org/doc/libs/1_45_0/libs/flyweight/doc/tutorial/index.html#guide)

Besides, I understand that clear flyweight pool is the minimal fix.
But since it's a more physical component than rc counter and supposed to be transparent to the core_worker, 
I don't think clearing the flyweight pool provides better semantics in here.

**Comment by jovany-wang** [comment]
Any update?

**Comment by jovany-wang** [comment]
@NKcqx Is there any update?

**Comment by jjyao** [comment]
Any updates?

**Comment by NKcqx** [comment]
Ooops, forget to check on the PR, so sorry for the delay.

As upgrading boost, I've encounter some bizarre problems: The compilation will crash at some mutex files when repeating compiling for about 10 times, and the CI flakey failure appears to be the same issue.

But I'm afraid I have to hold this PR for weeks as I'm recently focusing on other issues.
@jjyao @jovany-wang

**Comment by jovany-wang** [comment]
@NKcqx Should we keep this open?

**Pull request closed without merging**
